### ISSUE TITLE
Error in pcgrr::list_to_df(pcgr_config$tumor_type) : it should be a list

### ISSUE THREAD

**Issue by vladsaveliev** [sigven/cpsr#2]
Almost getting there with CPSR :) Errors in the end, not sure again if has to do with no-docker installation. A toy example:

```
python cpsr/cpsr.py --input_vcf cpsr/example.vcf.gz . cpsr grch37 cpsr/cpsr.toml example --no-docker

...

2018-10-30 23:10:54 - cpsr-gene-annotate - INFO - Number of PASSed variant calls: 1272
_frozen_importlib:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
2018-10-30 23:10:55 - cpsr-summarise - INFO - Converting VCF to TSV with https://github.com/sigven/vcf2tsv
_frozen_importlib:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
2018-10-30 23:10:59 - cpsr-summarise - INFO - Finished

2018-10-30 23:10:59 - cpsr-writer - INFO - STEP 4: Generation of output files - Cancer predisposition sequencing report
.libPaths():
/data/cephfs/punim0010/extras/vlad/miniconda/envs/pcgr/lib/R/library

Error in pcgrr::list_to_df(pcgr_config$tumor_type) : it should be a list
Calls: <Anonymous> -> <Anonymous> -> %>% -> eval -> eval -> <Anonymous>
Execution halted
```

/cc @pdiakumis

**Comment by sigven** [comment]
Hmm.. looking at it now. Thought you'd never reach this line when you initialize the CPSR report object (i.e. `type = 'predisposition'`). I'll have a closer look.

**Comment by sigven** [comment]
The numpy RuntimeWarning's I have seen before, I guess this might actually be no-docker related. Just re-ran the example VCF using Docker, and it did not produce the `pcgrr::list_to_df` error. Do not have a great clue what's going on there at the moment.

**Comment by vladsaveliev** [comment]
Good, thanks for looking into this! Might have to do with different package versions in image vs conda. Will look into this and pin all packages for good.

Just to make sure, the dev image and git master branch are on sync at the moment in terms of pcgrr and python soruce, right?

**Comment by pdiakumis** [comment]
Vlad and Sigve,
I think the error has something to do with a couple of things. Let's follow the crumbs:

* `pcgr/src/cpsr.R` calls `pcgrr::generate_predisposition_report`, which sort of generates the error you got:

```
> pcgrr::generate_predisposition_report(dir, query_vcf2tsv, pcgr_data, cpsr_config, sample_name, version, genome_assembly = genome_assembly)
Error in pcgrr::list_to_df(pcgr_config[["tier_model"]]) :                                                                                                                                                            
  it should be a list
```

* Inside `pcgrr::generate_predisposition_report` the function called is `pcgrr::get_calls`, which errors out and warns us because the `INTOGEN_DRIVER_MUT` column that's supposed to be in the tsv file is actually named `INTOGEN_DRIVER`:

```
> calls <- pcgrr::get_calls(query_vcf2tsv, pcgr_data, pcgr_version, sample_name, pcgr_config, genome_seq, assembly)                                                                                          

2018-11-04 15:51:47 [INFO] Excluding 0 variants from non-nuclear chromosomes/scaffolds                                                                                                                               
2018-11-04 15:51:48 [INFO] Number of PASS variants: 1272                                                                                                                                                             
2018-11-04 15:51:48 [INFO] Number of SNVs: 1222                                                                                                                                                                      
2018-11-04 15:51:48 [INFO] Number of deletions: 24                                                                                                                                                                   
2018-11-04 15:51:48 [INFO] Number of insertions: 0                                                                                                                                                                   
2018-11-04 15:51:48 [INFO] Number of block substitutions: 26                                                                                                                                                         
Error in `$<-.data.frame`(`*tmp*`, LOSS_OF_FUNCTION, value = TRUE) :                                                                                                                                                 
  replacement has 1 row, data has 0                                                                                                                                                                                  
In addition: Warning messages:                                                                                                                                                                                       
1: In is.na(vcf_data_df$INTOGEN_DRIVER_MUT) :                                                                                                                                                                        
  is.na() applied to non-(list or vector) of type 'NULL'                                                                                                                                                             
2: In is.na(vcf_data_df$INTOGEN_DRIVER_MUT) :                                                                                                                                                                        
  is.na() applied to non-(list or vector) of type 'NULL'                                                                                                                                                             
```

* Inside `pcgrr::get_calls` the tsv file gets read, and further down the wrong column name is called:

```
> vcf_data_df <- read.table(gzfile(tsv_gz_file), skip = 1, sep = "\t", header = T, stringsAsFactors = F, quote = "", comment.char = "", na.strings = c("."))
> colnames(vcf_data_df) %>% sort()
[.....]
 [76] "HGVSp_short"                  "HIGH_INF_POS"                 "ID"                           "IMPACT"                       "INTOGEN_DRIVER"
 [81] "INTRON"                       "LoF"                          "LoF_filter"                   "LoF_flags"                    "LoF_info"
[.....]

> [...]

>   if (nrow(vcf_data_df[!is.na(vcf_data_df$INTOGEN_DRIVER_MUT), ]) > 0){
+     vcf_data_df[!is.na(vcf_data_df$INTOGEN_DRIVER_MUT), ]$INTOGEN_DRIVER_MUT <- TRUE
+   }
Warning message:
In is.na(vcf_data_df$INTOGEN_DRIVER_MUT) :
  is.na() applied to non-(list or vector) of type 'NULL'
```

* But that's not why the program is crashing. Further down, we have:

```
>   if ("LoF" %in% colnames(vcf_data_df)){ # TRUE
+     vcf_data_df$LOSS_OF_FUNCTION <- FALSE # Create a LOSS_OF_FUNCTION column with all FALSEs
+     if(nrow(vcf_data_df[!is.na(vcf_data_df$LoF),]) > 0){ # Okay, there are 18 rows where this expression holds
+       vcf_data_df[!is.na(vcf_data_df$LoF) & vcf_data_df$LoF == "HC", ]$LOSS_OF_FUNCTION <- TRUE # But there are 0 rows where this expression holds
+     }
Error in `$<-.data.frame`(`*tmp*`, LOSS_OF_FUNCTION, value = TRUE) :
  replacement has 1 row, data has 0

> table(vcf_data_df$LoF, useNA="ifany") # No 'HC' here

  LC <NA>
  18 1254
```

This is because there are no rows in `vcf_data_df` that fit the expression used. Perhaps in the first expression above you should add the HC check, which would ensure the function would get skipped when empty.

I tried running the R script after correcting the two above and it generates the report fine for me at least.
Hope this helps a bit.

**Comment by vladsaveliev** [comment]
The LoF error hints that it might have to do with differences in LoFtee annotations (PCGR have an older version of the plugin packaged than conda's ensemble VEP), will explore that!

**Comment by sigven** [comment]
Peter and Vlad,
Thanks a lot for figuring this out! All the variable naming (`INTOGEN_DRIVER`, `INTOGEN_DRIVER_MUT` etc) should likely have been set up more robustly, but for now it has been working fairly ok. A clarification regarding this:

`pcgrr::get_calls`is a function that takes the vcf2tsv-converted, annotated VCF file and adds additional annotations. This is done both for the annotated VCF that comes with the PCGR workflow, and also the annotated VCF that is produced by the CPSR workflow. For PCGR, the VCF is also annotated with somatic driver mutations (i.e. `pcgr_vcfanno .. --intogen_driver_mut`), while this is not done for CPSR where the focus is solely germline. Implying that `INTOGEN_DRIVER_MUT`will be present in PCGR, but not in CPSR. 

The variable `INTOGEN_DRIVER` is on the other hand an annotation at the gene level, denoting a particular gene as a predicted cancer driver gene (from the IntoGen resource), and this will come both with the PCGR and CPSR, since all gene annotations are similar for the two workflows. 

Basically, I need to do a proper check for the presence of the `INTOGEN_DRIVER_MUT` column in the `pcgrr::get_calls` function. 

Hope this may clarify somewhat. Either way very grateful that you highlighted this weak point:-) More later.

Sigve

**Comment by vladsaveliev** [comment]
Sigve, thanks so much for looking into it!

That would be good to have it more robust for the future, but first I tried to figure out why the same test works differently with dockerized and codnarized version. Got to the bottom of the missing `HC` bug: it stems from different versions of loftee coming from the image vs. conda. Now I changed condarized version to reuse your built-in `loftee.tgz` and it helped: `.tsv` files that are passed into `cpsr.R` are identical in both installations.

However, I'm still having the original error with condarized version for some reason - digging into it now:

```
Error in pcgrr::list_to_df(pcgr_config$tumor_type) : it should be a list
Calls: <Anonymous> -> <Anonymous> -> %>% -> eval -> eval -> <Anonymous>
Execution halted
```
